src/enumerable.cr

```diff
@@ -1419,4 +1419,223 @@ module Enumerable(T)
       hash[key] = value
     end
   end
+
+  # Yields elements of `self` and *others* in tandem to the given block.
+  #
+  # Raises an `IndexError` if any of *others* doesn't have as many elements
+  # as `self`. See `zip?` for a version that yields `nil` instead of raising.
+  #
+  # ```
+  # a = [1, 2, 3]
+  # b = ["a", "b", "c"]
+  #
+  # a.zip(b) { |x, y| puts "#{x} -- #{y}" }
+  # ```
+  #
+  # The above produces:
+  #
+  # ```text
+  # 1 -- a
+  # 2 -- b
+  # 3 -- c
+  # ```
+  #
+  # An example with multiple arguments:
+  #
+  # ```
+  # (1..3).zip(4..6, 7..9) do |x, y, z|
+  #   puts "#{x} -- #{y} -- #{z}"
+  # end
+  # ```
+  #
+  # The above produces:
+  #
+  # ```text
+  # 1 -- 4 -- 7
+  # 2 -- 5 -- 8
+  # 3 -- 6 -- 9
+  # ```
+  def zip(*others : Indexable | Iterable | Iterator, &block)
+    Enumerable.zip(self, others) do |elems|
+      yield elems
+    end
+  end
+
+  # Returns an `Array` of tuples populated with the elements of `self` and
+  # *others* traversed in tandem.
+  #
+  # Raises an `IndexError` if any of *others* doesn't have as many elements
+  # as `self`. See `zip?` for a version that yields `nil` instead of raising.
+  #
+  # ```
+  # a = [1, 2, 3]
+  # b = ["a", "b", "c"]
+  #
+  # a.zip(b) # => [{1, "a"}, {2, "b"}, {3, "c"}]
+  # ```
+  #
+  # An example with multiple arguments:
+  #
+  # ```
+  # a = [1, 2, 3]
+  # b = (4..6)
+  # c = 8.downto(3)
+  #
+  # a.zip(b, c) # => [{1, 4, 8}, {2, 5, 7}, {3, 6, 6}]
+  # ```
+  def zip(*others : Indexable | Iterable | Iterator)
+    pairs = Array(typeof(zip(*others) { |e| break e }.not_nil!)).new(size)
```

Comment:
How does that type resolution handle a `other` collection whose elements are nilable?

Reply:
It's the tuple that becomes nilable because the compiler can't know whether the block will be called (or something like that). This doesn't affect individual elements.